src/main/java/uk/gov/gchq/magmacore/service/DbChangeSet.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package uk.gov.gchq.magmacore.service;
+
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Class representing an invertible set of deletes and creates.
+ */
+public class DbChangeSet implements Function<MagmaCoreService, MagmaCoreService> {
+    private Set<DbDeleteOperation> deletes;
+    private Set<DbCreateOperation> creates;
+
+    /**
+     * Constructor.
+     *
+     * @param deletes a {@link Set} of {@link DbDeleteOperation}
+     * @param creates a {@link Set} of {@link DbCreateOperation}
+    */
+    public DbChangeSet(final Set<DbDeleteOperation> deletes, final Set<DbCreateOperation> creates) {
+        this.deletes = deletes;
+        this.creates = creates;
+    }
+
+    /**
+     * Apply this {@link DbChangeSet} to a {@link MagmaCoreService}.
+     *
+     * */
+    @Override
+    public MagmaCoreService apply(final MagmaCoreService mcService) {
+        final var deleteFunction = deletes
+            .stream()
+            .map(x -> (Function<MagmaCoreService, MagmaCoreService>) x)
+            .reduce(Function::andThen)
+            .orElse(Function.identity());
+
+        final var createFunction = creates
+            .stream()
+            .map(x -> (Function<MagmaCoreService, MagmaCoreService>) x)
+            .reduce(Function::andThen)
+            .orElse(Function.identity());
+
+        mcService.runInTransaction(deleteFunction.andThen(createFunction));
+        return mcService;
+    }
+
+    /**
+     * Invert a {@link DbChangeSet}.
+     *
+     * @param c a {@link DbChangeSet}
+     * @return a {@link DbChangeSet}
+    */
+    public static DbChangeSet invert(final DbChangeSet c) {
```

Comment:
No - some functions can be inverted and some can't. If you apply a function and then apply its inverse the effect is a no-op (aka an identity operation), so while the effect of _applying_ the inverse does revert a previously _applied_ function, this is just inverting the function rather than applying it, so nothing is actually reverted until is is applied some time later.

Reply:
Thank you for clarifying 😄